server/odc-service/src/main/java/com/oceanbase/odc/service/session/model/AsyncExecuteContext.java

```diff
@@ -0,0 +1,140 @@
+import com.oceanbase.odc.service.iam.model.User;
+import com.oceanbase.odc.service.iam.util.SecurityContextUtils;
+
+import lombok.Getter;
+import lombok.Setter;
+import lombok.extern.slf4j.Slf4j;
+
+/**
+ * @author liuyizhuo.lyz
+ * @date 2024/4/15
+ */
+@Getter
+@Slf4j
+public class AsyncExecuteContext<T> {
+    public static final String SHOW_TABLE_COLUMN_INFO = "SHOW_TABLE_COLUMN_INFO";
+
+    private final Function<JdbcGeneralResult, T> mapper;
+    private final ConnectionSession session;
+    private final List<SqlTuple> sqlTuples;
+    private final Queue<T> results = new ConcurrentLinkedQueue<>();
+    private final List<BiConsumer<ConnectionSession, String>> traceIdHooks = new ArrayList<>();
+    private final User user;
+    private final Long waitTimeMillis;
+    private final List<String> sessionIds;
+
+    @Setter
+    private Future<List<JdbcGeneralResult>> future;
+    private String currentTraceId;
+    @Setter
+    private String sql;
+    private int count = 0;
+
+    public AsyncExecuteContext(ConnectionSession session, List<SqlTuple> sqlTuples,
+            Function<JdbcGeneralResult, T> mapper, User user, Long waitTimeMillis) {
+        this.session = session;
+        this.sqlTuples = sqlTuples;
+        this.mapper = mapper;
+        this.user = user;
+        this.waitTimeMillis = waitTimeMillis;
+        this.sessionIds = initSessionIds(session);
+    }
+
+    public int getTotal() {
+        return sqlTuples.size();
+    }
+
+    public List<T> getResults() {
+        List<T> copiedResults = new ArrayList<>();
+        while (!results.isEmpty()) {
+            copiedResults.add(results.poll());
+        }
+        return copiedResults;
+    }
+
+    public void setCurrentTraceId(String traceId) {
+        currentTraceId = traceId;
+        for (BiConsumer<ConnectionSession, String> hook : traceIdHooks) {
+            hook.accept(session, traceId);
+        }
+    }
+
+    public void onQueryStart(String sql) {
+        count++;
+        this.sql = sql;
+    }
+
+    public void onQueryExecuting() {
+        String traceId = session.getSyncJdbcExecutor(BACKEND_DS_KEY)
+                .execute((StatementCallback<String>) stmt -> OBUtils
+                        .queryTraceIdFromASH(stmt, sessionIds, session.getConnectType()));
+        if (traceId != null) {
+            setCurrentTraceId(traceId);
+        }
+    }
+
+    public void onQueryFinish(List<JdbcGeneralResult> results) {
+        SecurityContextUtils.setCurrentUser(user);
+        try {
+            for (JdbcGeneralResult result : results) {
+                this.results.add(mapper.apply(result));
+                for (BiConsumer<ConnectionSession, String> hook : traceIdHooks) {
+                    hook.accept(session, result.getTraceId());
+                }
+            }
+        } finally {
+            SecurityContextUtils.clear();
+        }
+    }
+
+    private List<String> initSessionIds(ConnectionSession session) {
```

Comment:
AsyncExecuteContext is a general component, we'd better not put in anything unique to OB.

Reply:
done